### ISSUE TITLE
Compiled inference plan returns wrong output tensor when forward ends in a metadata view (Reshape/Squeeze/etc.)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#228]
## Summary

`CompiledModelCache.GetOrCompileInference` (and the underlying `CompiledInferencePlan.Compile`) silently pick the **wrong output tensor** whenever the user's forward lambda ends with a pure metadata-view op (`Reshape` on contiguous data, `Squeeze`, `Unsqueeze`, etc.) or contains host-side control flow that conditionally appends such views.

The replayed `plan.Execute()` returns an **earlier** intermediate tensor rather than what the caller returned. There is no warning, no exception — just wrong data with a different shape.

Downstream, this corrupted 11 neural-network classes in [ooples/AiDotNet](https://github.com/ooples/AiDotNet) (ResNet, VGG, EfficientNet, MobileNetV2, CNN, UNet3D, VoxelCNN, FastText, GloVe, Word2Vec, SiameseNeuralNetwork) when AiDotNet PR #1155 started routing their `Predict` through the compiled path. A ResNet-18 `[3, 32, 32] -> logits [10]` inference returns a `[64]`-shaped feature map instead. The AiDotNet hotfix ([PR #1167](https://github.com/ooples/AiDotNet/pull/1167)) restores the `Predict -> Forward` override for the affected models so they bypass the compiled path entirely — that is a band-aid, not a fix. The real bug is here.

## Minimal standalone repro (Tensors only)

Paste this file into `tests/AiDotNet.Tensors.Tests/Engines/Compilation/CompiledPlanFinalOutputReproTest.cs`:

```csharp
using AiDotNet.Tensors.Engines;
using AiDotNet.Tensors.Engines.Compilation;
using AiDotNet.Tensors.LinearAlgebra;
using Xunit;
using Xunit.Abstractions;

namespace AiDotNet.Tensors.Tests.Engines.Compilation;

public class CompiledPlanFinalOutputReproTest
{
    private readonly ITestOutputHelper _output;
    public CompiledPlanFinalOutputReproTest(ITestOutputHelper output) { _output = output; }

    /// Forward ends in a Reshape view of an intermediate tensor.
    /// Reshape on contiguous data does NOT record a lazy node
    /// (Tensor.cs:1509-1527 returns a fresh Tensor<T> via the view
    /// constructor and never touches LazyTensorScope.Record*), so the
    /// compiled plan picks the last pre-reshape op's output buffer and
    /// replay returns its shape instead of the reshape's shape.
    [Fact]
    public void CompiledPlan_ForwardReturnsReshape_ReturnsWrongShape()
    {
        using var cache = new CompiledModelCache<float>();
        var engine = new CpuEngine();

        var input = new Tensor<float>(new[] { 1, 6 });
        var weights = new Tensor<float>(new[] { 6, 10 });
        for (int i = 0; i < input.Length; i++) input[i] = 1.0f;
        for (int i = 0; i < weights.Length; i++) weights[i] = 0.1f;

        var expectedEager = engine.TensorMatMul(input, weights).Reshape(new[] { 10 });
        Assert.Equal(new[] { 10 }, expectedEager.Shape.ToArray());

        var plan = cache.GetOrCompileInference(input._shape, () =>
        {
            var matmul = engine.TensorMatMul(input, weights);   // lazy node
            var result = matmul.Reshape(new[] { 10 });          // metadata view — NOT a lazy node
            // Return value is discarded because the API type is Action.
        });

        var compiledOutput = plan.Execute();
        Assert.Equal(new[] { 10 }, compiledOutput.Shape.ToArray());
    }

    /// Host-side branch — the exact pattern in AiDotNet ResNet / VGG / CNN
    /// forwards (promote rank-3 input to rank-4, then optionally strip the
    /// synthetic batch dim with Reshape).
    [Fact]
    public void CompiledPlan_ForwardWithHostSideBranch_ReturnsWrongShape()
    {
        using var cache = new CompiledModelCache<float>();
        var engine = new CpuEngine();

        var input = new Tensor<float>(new[] { 1, 6 });
        var weights = new Tensor<float>(new[] { 6, 10 });
        for (int i = 0; i < input.Length; i++) input[i] = 1.0f;
        for (int i = 0; i < weights.Length; i++) weights[i] = 0.1f;

        var plan = cache.GetOrCompileInference(input._shape, () =>
        {
            var output = engine.TensorMatMul(input, weights);
            if (output.Rank == 2 && output.Shape[0] == 1)
                output = output.Reshape(new[] { output.Shape[1] });
        });

        var compiledOutput = plan.Execute();
        Assert.Equal(new[] { 10 }, compiledOutput.Shape.ToArray());
    }

    /// Canary: Transpose ends the forward. TensorTranspose goes through
    /// LazyTensorScope.Record* so a proper lazy node IS created, and the
    /// "last-node heuristic" happens to match the caller's intent here.
    /// Passing this while the two above fail isolates the bug to pure
    /// metadata views, not to lazy nodes generally.
    [Fact]
    public void CompiledPlan_ForwardReturnsTranspose_ReturnsCorrectShape()
    {
        using var cache = new CompiledModelCache<float>();
        var engine = new CpuEngine();

        var input = new Tensor<float>(new[] { 4, 6 });
        var weights = new Tensor<float>(new[] { 6, 10 });
        for (int i = 0; i < input.Length; i++) input[i] = 1.0f;
        for (int i = 0; i < weights.Length; i++) weights[i] = 0.1f;

        var plan = cache.GetOrCompileInference(input._shape, () =>
        {
            var mm = engine.TensorMatMul(input, weights);
            var t = engine.TensorTranspose(mm);
        });

        var compiledOutput = plan.Execute();
        Assert.Equal(new[] { 10, 4 }, compiledOutput.Shape.ToArray());
    }
}
```

### Observed on `main` (this repo, net10.0, CpuEngine)

```
AiDotNet.Tensors.Tests.Engines.Compilation.CompiledPlanFinalOutputReproTest
    CompiledPlan_ForwardReturnsReshape_ReturnsWrongShape          [FAIL]
        Expected: [10]   Actual: [1, 10]
    CompiledPlan_ForwardWithHostSideBranch_ReturnsWrongShape      [FAIL]
        Expected: [10]   Actual: [1, 10]
    CompiledPlan_ForwardReturnsTranspose_ReturnsCorrectShape      [PASS]
```

## Root cause

Two defects that compound:

**(1)** `GetOrCompileInference` takes `Action`, not `Func<Tensor<T>>`.

`src/AiDotNet.Tensors/Engines/Compilation/CompiledModelCache.cs`:

```csharp
// line 36
public ICompiledPlan<T> GetOrCompileInference(int[] inputShape, Action forward)
// line 73
public ICompiledPlan<T> GetOrCompileInference(Tensor<T> input, Action forward)
// line 176
public ICompiledPlan<T> GetOrCompileInference(int[] inputShape, Action forward, SymbolicShape symbolicShape)
```

The caller's `Func<Tensor<T>>` is implicitly coerced to `Action` and its return value is dropped. There is no channel for the caller to tell the compiler *which* tensor is the output.

**(2)** The plan then guesses, and the guess is the last optimized step.

`src/AiDotNet.Tensors/Engines/Compilation/CompiledInferencePlan.cs:644-645`:

```csharp
// Use the last optimized step's output (may differ from original after fusion/spectral passes)
var finalOutput = optimizedSteps.Length > 0
    ? optimizedSteps[optimizedSteps.Length - 1].OutputBuffer
    : new Tensor<T>(new int[] { 0 });
```

The inline comment already admits this is fragile. The heuristic breaks whenever the caller's returned tensor is *not* produced by the topologically-last `LazyNode`, which is the case for:

- `Reshape(...)` on a contiguous tensor (`src/AiDotNet.Tensors/LinearAlgebra/Tensor.cs:1509-1527`) — returns `new Tensor<T>(_data, newShape, newStrides, _storageOffset, _storage)` with no `LazySource` assignment, so no node is recorded.
- Any other "view" op that only touches `_shape`/`_strides`/`_storageOffset` without going through `LazyTensorScope.RecordUnary/RecordBinary/RecordVariadic`.
- A `forward` that conditionally skips its trailing ops based on host-side state (`tensor.Rank`, `tensor.Shape[i]`, caller flags), leaving the intended output upstream of the last recorded node.
- Any optimization pass that reorders or splits steps such that the caller-visible output is no longer `steps[^1]`.

`Execute()` then returns `_finalOutput` — not the caller's tensor — silently:

`src/AiDotNet.Tensors/Engines/Compilation/CompiledInferencePlan.cs:443`:

```csharp
return _finalOutput;
```

## Why the AiDotNet hotfix (PR #1167) is not a real fix

PR #1167 restores `public override Tensor<T> Predict(Tensor<T> input) => Forward(input)` in 11 model classes so `Predict` bypasses the compiled path. That works for those models today but:

- Any *new* model with a custom `Forward` (shape-conditional control flow, metadata views in the tail) hits the same bug as soon as it lands without the explicit override.
- Every user who calls `CompileForward(sampleInput)` directly on those models still trips it.
- Every consumer of the Tensors library who uses `CompiledModelCache.GetOrCompileInference` with a forward that returns a view hits it.
- The "Subclasses whose inference is just a flat forward pass through Layers should leave the default in place to pick up compiled replay automatically" contract in `NeuralNetworkBase.Predict`'s docstring is now a lie until this is fixed.

The band-aid needs to come off when this issue is resolved: revert PR #1167's overrides and let compiled replay take over once it produces the caller's actual output.

## Proposed fix

Three changes, minimal surface:

1. **Add `Func<Tensor<T>>` overloads** on `CompiledModelCache` alongside the existing `Action` ones:

   ```csharp
   public ICompiledPlan<T> GetOrCompileInference(int[] inputShape, Func<Tensor<T>> forward);
   public ICompiledPlan<T> GetOrCompileInference(Tensor<T> input, Func<Tensor<T>> forward);
   public ICompiledPlan<T> GetOrCompileInference(int[] inputShape, Func<Tensor<T>> forward, SymbolicShape symbolicShape);
   ```

   Keep the `Action` overloads for source compat; mark them `[Obsolete("Use Func<Tensor<T>> overload; Action drops the output tensor identity — see issue #XXX")]`.

2. **Propagate the returned tensor through compile:**

   - Capture `outputTensor = forward()` inside the `GraphMode.Enable()` scope.
   - Pass it to `LazyTensorScope.CompileInference<T>(Tensor<T> explicitOutput)`.
   - Thread it through to `CompiledInferencePlan<T>.Compile(scope, engine, explicitOutput)`.
   - Use `explicitOutput` as `_finalOutput` instead of the last-step heuristic.

3. **Handle the "output is a view with no LazySource" case** (contiguous `Reshape`, `Transpose` on contiguous, etc.): when `explicitOutput.LazySource == null` but the tensor's `_storage` matches a lazy node's output buffer, the plan knows the view needs to be reconstructed from that producer's storage after each `Execute()`. Options:

   (a) Record `Reshape` / view ops as a lightweight lazy node (preferred — keeps the graph complete; cheapest implementation is a no-op compute step that just rebinds shape/strides).

   (b) Store the *reconstruction recipe* (`{producer node, newShape, newStrides, offset}`) alongside `_finalOutput` and reapply it inside `Execute()` before returning.

   Option (a) is less code and keeps `ExecuteInto` / `SetInputs` / plan stitching working uniformly.

## Acceptance criteria

- All three repro tests above pass on `main` with both `EnableCompilation=true` and optimization passes enabled.
- Adding the same test against `TensorPermute`, `TensorSqueeze`, `TensorUnsqueeze`, and `TensorContiguous` tails also passes (same class of bug).
- The ResNet-18 / VGG / EfficientNet / MobileNetV2 / CNN / UNet3D / VoxelCNN / FastText / GloVe / Word2Vec / SiameseNeuralNetwork `Predict` tests in ooples/AiDotNet pass with the PR #1167 `Predict` overrides reverted.
- No regression in the existing `CompilationComponentTests`, `PlanStitchingTests`, `ExecuteIntoTests`, `MultiDimSymbolicShapeTests`, `InferencePlanSerializationTests`.
- Serialization round-trip (`InferencePlanWriter.Write/Read`) preserves the explicit output identity — a plan deserialized from disk should `Execute()` to the same shape it compiled to.

## Workaround (what downstream is doing today)

[ooples/AiDotNet PR #1167](https://github.com/ooples/AiDotNet/pull/1167) — restore `public override Tensor<T> Predict(Tensor<T> input) => Forward(input)` in every model whose `Forward` is not a flat layer pipeline. Track it as the revert-me sibling of this issue.

## References

- Downstream repro + diagnostic:
  `Forward direct shape: [10]; Predict compile-off shape: [10]; Predict compile-on shape: [64]`
  on ResNet-18 / `[3, 32, 32]` / `numClasses=10` — see PR #1167 description.
- AiDotNet PR that exposed the bug: [ooples/AiDotNet#1155](https://github.com/ooples/AiDotNet/pull/1155) (now merged, `chore: remove dead AiDotNet JIT scaffolding; bump Tensors to 0.42.3`).
- Hotfix masking the symptom: [ooples/AiDotNet#1167](https://github.com/ooples/AiDotNet/pull/1167).
